Fix/(llms.txt) not crawling links inside of file

Chillbruhhh opened the pull request:
# Pull Request

## Summary
<!-- Provide a brief description of what this PR accomplishes -->

I discovered archon wasn't properly crawling llms.txt, llms.md etc. and i added the feature in to where it will now parse and crawl the links in the llms.txt file, its backwards compatible and will crawl llms.txt even if they don't have links, just fixes that bug. 

This is what it was crawling before when crawling a llms.txt:

<img width="1636" height="1195" alt="Screenshot 2025-08-22 035117" src="https://github.com/user-attachments/assets/72ff31be-d6a5-471a-a843-f898c192037a" />


This is what it looks like crawling llms.txt now: 

<img width="1483" height="1178" alt="Screenshot 2025-08-22 041311" src="https://github.com/user-attachments/assets/19effe63-6714-429c-9a90-528ec84ba2bf" />

<img width="487" height="250" alt="Screenshot 2025-08-22 041337" src="https://github.com/user-attachments/assets/db9854fa-07b3-4f00-aab6-8a963b5c1a49" />



## Changes Made
<!-- List the main changes in this PR -->
 Enhanced llms.txt support - The system now automatically detects and crawls all links found inside llms.txt files, instead of just treating them as static text files.

 ## Key Changes Made

  1. Added Link Detection & Extraction (url_handler.py)

  - extract_markdown_links() - Extracts [text](url) patterns from file content
  - is_link_collection_file() - Detects llms.txt files by name and content analysis
  - Handles relative URLs, filters invalid links, removes duplicates

  2. Enhanced Crawling Logic (crawling_service.py)

  - Before: Text files → Single document
  - After: Text files → Check if link collection → Extract links → Batch crawl all links → Combine results

## Type of Change
<!-- Mark the relevant option with an "x" -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Performance improvement
- [x] Code refactoring

## Affected Services
<!-- Mark all that apply with an "x" -->
- [x] Frontend (React UI)
- [x] Server (FastAPI backend)

## Testing
<!-- Describe how you tested your changes -->
- [x] All existing tests pass
- [x] Manually tested affected user flows
- [x] Docker builds succeed for all services

### Test Evidence
<!-- Provide specific test commands run and their results -->
```bash
 1. Testing Ultimate URL Format Support:
     Extracted 13 unique links from content
        Found 13 URLs (expected: 10+):
        - https://platform.openai.com/docs
        - https://docs.anthropic.com/
        - https://python.langchain.com/docs
        - https://www.langchain.com/langsmith
        - https://wandb.ai/
        - https://example.com/
        - https://github.com/microsoft/vscode
        - https://raw.githubusercontent.com/microsoft/vscode/main/README.md
        - https://www.google.com
        - https://www.stackoverflow.com/questions
        - https://example.com/test
        - https://www.example.com
        - https://test.com/path?query=1#fragment

     2. Testing Enhanced Markdown File Detection:
     ✓ Detected link collection file by filename: llms.txt
        https://example.com/llms.txt
          Markdown: ✗
          Collection: ✓
     ✓ Detected link collection file by filename: llms.md
        https://github.com/user/repo/llms.md
          Markdown: ✓
          Collection: ✓
     ✓ Detected link collection file by filename: links.mdx
        https://example.com/links.mdx
          Markdown: ✓
          Collection: ✓
     ✓ Detected link collection file by filename: resources.markdown
        https://example.com/resources.markdown
          Markdown: ✓
          Collection: ✓
     ✓ Detected link collection file by filename: references.txt
        https://example.com/references.txt
          Markdown: ✗
          Collection: ✓
     ✓ Detected potential link collection file: awesome-references.md
        https://example.com/awesome-references.md
          Markdown: ✓
          Collection: ✓
        https://example.com/regular-file.py
          Markdown: ✗
          Collection: ✗

     3. Testing DRY Principle Implementation:
     Extracted 7 unique links from content
     ✓ Detected link collection by content analysis: 7 links, density 4.67%
        DRY content analysis: ✓ (should reuse main extractor)

     4. Testing GitHub URL Enhancement:
     Transformed GitHub file URL to raw: https://github.com/microsoft/vscode/blob/main/README.md -> https://raw.githubusercontent.com/microsoft/vscode/main/README.md
        Original:    https://github.com/microsoft/vscode/blob/main/README.md
        Transformed: https://raw.githubusercontent.com/microsoft/vscode/main/README.md

     Transformed GitHub file URL to raw: https://github.com/microsoft/vscode/blob/main/README.md?plain=1 -> https://raw.githubusercontent.com/microsoft/vscode/main/README.md
        Original:    https://github.com/microsoft/vscode/blob/main/README.md?plain=1
        Transformed: https://raw.githubusercontent.com/microsoft/vscode/main/README.md

     Transformed GitHub file URL to raw: https://github.com/microsoft/vscode/blob/main/README.md#installation -> https://raw.githubusercontent.com/microsoft/vscode/main/README.md
        Original:    https://github.com/microsoft/vscode/blob/main/README.md#installation
        Transformed: https://raw.githubusercontent.com/microsoft/vscode/main/README.md

     Transformed GitHub file URL to raw: https://github.com/microsoft/vscode/blob/main/README.md?plain=1#installation -> https://raw.githubusercontent.com/microsoft/vscode/main/README.md
        Original:    https://github.com/microsoft/vscode/blob/main/README.md?plain=1#installation
        Transformed: https://raw.githubusercontent.com/microsoft/vscode/main/README.md

     ================================================================================

     🛡️  CodeRabbit-Proof Features Implemented:
       ✅ Named regex groups with ultimate URL format support
       ✅ www.example.com and //example.com detection
       ✅ Enhanced punctuation cleanup (.,;:)]>)
       ✅ DRY principle - single source of truth for patterns
       ✅ Bulletproof GitHub URL handling with query/fragment stripping
       ✅ Complete pattern coverage including 'references'
       ✅ Database configuration respect for concurrency
       ✅ Markdown file support (.md, .mdx, .markdown)

#trust me it works
```

## Checklist
<!-- Mark completed items with an "x" -->
- [x] My code follows the service architecture patterns
- [x] If using an AI coding assistant, I used the CLAUDE.md rules
- [x] All new and existing tests pass locally
- [x] My changes generate no new warnings
- [x] I have verified no regressions in existing features

## Breaking Changes
<!-- If this PR introduces breaking changes, describe them here -->
<!-- Include migration steps if applicable -->

## Additional Notes
<!-- Any additional information that reviewers should know -->
<!-- Screenshots, performance metrics, dependencies added, etc. -->

<img width="431" height="833" alt="Screenshot 2025-08-22 031955" src="https://github.com/user-attachments/assets/657a7ae6-6767-4ea7-9449-d5d464a1b404" />


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* New Features
  * Markdown files are now crawled like text, with automatic detection of “link collection” docs that extract, filter, and batch-crawl embedded links.
  * Multi-stage crawling for Markdown/text with clearer, threshold-based progress updates.
  * Stage-specific progress indicators for documents and code (current/total batches) for more accurate status tracking.
  * Cleaner, user-friendly display names derived from URLs.

* Bug Fixes
  * More reliable GitHub file URL handling that correctly resolves raw file links.
  * Improved filtering to avoid self-referential and binary links when processing link collections.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Chillbruhhh commented:
ready for merge

coleam00 commented:
Thank you for this @Chillbruhhh! I tested just now and this works beautifully for llms.txt that have a lot of links within.

However, I tested with an llms-full.txt (https://ai.pydantic.dev/llms-full.txt) and this used to be really fast since it was just a single page but now with this PR it found 492 URLs and it's going through all of those when it shouldn't have to. It's just a single page.

Also the progress for crawling pages is always at 100%:
<img width="944" height="829" alt="PydanticAIDocs100%" src="https://github.com/user-attachments/assets/a8a09f71-79ba-44ea-a969-c954e837c6bf" />

If you address these two things I think the PR would be ready! Probably good to rebase from main too.

Chillbruhhh commented:
@coleam00 

ready for merge! fixed the full-llms.txt issue. fixed the ui web crawling progress display! rebased it! 

<img width="434" height="582" alt="Screenshot 2025-08-30 213750" src="https://github.com/user-attachments/assets/7cdf57b5-44fc-43c7-bb0b-95e447a4c468" />

also made a change in the UI when storing batches, noticed sometimes, not all the time but sometimes the progress bar would break 
<img width="526" height="366" alt="Screenshot 2025-08-30 215858" src="https://github.com/user-attachments/assets/7d39e7b9-9f9a-43b5-9dca-e7d13e284bc3" />

Prevented this and fixed the other issues! checked with code rabbit i my ide aswell and it said good to go , no nitpick nothing. curious to see what it will say here in the PR.

GioPetro commented:
We need this feature !!! Please proceed

coleam00 commented:
@Chillbruhhh This PR is working really nicely for llms.txt - awesome work!

However, for llms-full.txt, it still isn't ideal. I tried with:

https://ai.pydantic.dev/llms-full.txt

And it says it's crawling 492 pages when it's really just 1, and on main right now it does as I'd expect - it quickly crawls the llms-full.txt as a single page and within seconds moves on to the embedding and storage step.
